File: Cpp/3 Sum.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    /**
     * @param numbers : Give an array numbers of n integer
     * @return : Find all unique triplets in the array which gives the sum of zero.
     */
    vector<vector<int> > threeSum(vector<int> &nums) {
        // write your code here
        sort(nums.begin(), nums.end());
        vector<vector<int> > vec;
        for (int i = 0; i + 2 < nums.size(); i++) {
            if (i > 0 && nums[i-1] == nums[i]) continue;
            int left = i + 1, right = nums.size() - 1;
            while (left < right) {
                int s = nums[left] + nums[right] + nums[i];
                if (s == 0) {
                    vector<int> res;
                    res.push_back(nums[i]);
                    res.push_back(nums[left]);
                    res.push_back(nums[right]);
                    vec.push_back(res);
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left-1]) {
                        left ++;
                    }
                    while (left < right && nums[right] == nums[right+1]) {
                        right --;
                    }
                } else if (s < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return vec;
    }
};

File: Java/3-Sum.java
public class Solution {
    /**
     * @param numbers : Give an array numbers of n integer
     * @return : Find all unique triplets in the array which gives the sum of zero.
     */
    public ArrayList<ArrayList<Integer>> threeSum(int[] numbers) {
        // write your code here
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        Arrays.sort(numbers);
        for (int i = 0; i + 2 < numbers.length; i ++) {
            if (i != 0 && numbers[i] == numbers[i - 1]) continue;
            int left = i + 1, right = numbers.length - 1;
            while (left < right) {
                int temp = numbers[i] + numbers[left] + numbers[right];
                if (temp == 0) {
                    ArrayList<Integer> cur = new ArrayList<Integer>();
                    cur.add(numbers[i]);
                    cur.add(numbers[left]);
                    cur.add(numbers[right]);
                    res.add(cur);
                    int leftv = numbers[left++], rightv = numbers[right--];
                    while (left < right && leftv == numbers[left]) left ++;
                    while (left < right && rightv == numbers[right]) right --;
                } else if(temp > 0) {
                    right --;
                } else {
                    left ++;
                }
            }
        }
        return res;
    }
}
